alf/algorithms/sac_algorithm.py

```diff
@@ -190,53 +194,80 @@ def predict(self, time_step: ActionTimeStep, state=None):
             network_state=state)
         return PolicyStep(action=action, state=state, info=())
 
-    def _actor_train_step(self, exp: Experience, state: SacActorState, action,
-                          log_pi):
-        critic_input = (exp.observation, action)
-
-        with tf.GradientTape(watch_accessed_variables=False) as tape:
-            tape.watch(action)
+    def _actor_train_step(self, exp: Experience, state: SacActorState,
+                          action_distribution, action, log_pi):
+
+        if self._is_continuous:
+            critic_input = (exp.observation, action)
+
+            with tf.GradientTape(watch_accessed_variables=False) as tape:
+                tape.watch(action)
+                critic1, critic1_state = self._critic_network1(
+                    critic_input,
+                    step_type=exp.step_type,
+                    network_state=state.critic1)
+
+                critic2, critic2_state = self._critic_network2(
+                    critic_input,
+                    step_type=exp.step_type,
+                    network_state=state.critic2)
+
+                target_q_value = tf.minimum(critic1, critic2)
+
+            dqda = tape.gradient(target_q_value, action)
+
+            def actor_loss_fn(dqda, action):
+                if self._dqda_clipping:
+                    dqda = tf.clip_by_value(dqda, -self._dqda_clipping,
+                                            self._dqda_clipping)
+                loss = 0.5 * losses.element_wise_squared_loss(
+                    tf.stop_gradient(dqda + action), action)
+                loss = tf.reduce_sum(
+                    loss, axis=list(range(1, len(loss.shape))))
+                return loss
+
+            actor_loss = tf.nest.map_structure(actor_loss_fn, dqda, action)
+            alpha = tf.stop_gradient(tf.exp(self._log_alpha))
+            actor_loss += alpha * log_pi
+        else:
             critic1, critic1_state = self._critic_network1(
-                critic_input,
+                exp.observation,
                 step_type=exp.step_type,
                 network_state=state.critic1)
 
             critic2, critic2_state = self._critic_network2(
-                critic_input,
+                exp.observation,
                 step_type=exp.step_type,
                 network_state=state.critic2)
 
-            target_q_value = tf.minimum(critic1, critic2)
-
-        dqda = tape.gradient(target_q_value, action)
-
-        def actor_loss_fn(dqda, action):
-            if self._dqda_clipping:
-                dqda = tf.clip_by_value(dqda, -self._dqda_clipping,
-                                        self._dqda_clipping)
-            loss = 0.5 * losses.element_wise_squared_loss(
-                tf.stop_gradient(dqda + action), action)
-            loss = tf.reduce_sum(loss, axis=list(range(1, len(loss.shape))))
-            return loss
+            action_probs = action_distribution._probs
```

Comment:
_probs seems to be a private member of distribution. We should avoid using it. Can you find some other ways to get it, like the extracting parameters for distribution when it’s stored in experience buffer